File: src/forager/ui/pages/gamepage.py
from __future__ import annotations
from PySide6.QtCore import Qt, Signal
from PySide6.QtGui import QPixmap, QFont, QColor
from PySide6.QtWidgets import (
    QWidget, QVBoxLayout, QHBoxLayout, QLabel, QPushButton, QScrollArea,
    QFrame,
)

from forager.core.game import Game
from forager.artwork import pipeline as art
from forager.ui.fonts import UI_FONT
from forager.ui.widgets.banner import Banner
from forager.ui.theme import C
from forager.ui.icons import load_icon as load_bundled_icon
from forager.ui import style

_PLAY_QSS = style.button_qss("play")
_RUNNING_QSS = style.button_qss("running")


class GamePage(QWidget):
    play = Signal(object)
    stop = Signal(object)
    back_requested = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self.game: Game | None = None
        self._logo: QPixmap | None = None
        self._running = False

        self.setStyleSheet(f"background-color: {C.COLOR_1};")

        outer = QVBoxLayout(self)
        outer.setContentsMargins(16, 16, 16, 16)
        outer.setSpacing(0)

        scroll = QScrollArea()
        scroll.setWidgetResizable(True)
        scroll.setFrameShape(QFrame.Shape.NoFrame)
        scroll.setStyleSheet(f"QScrollArea {{ background: {C.COLOR_1}; border: none; }}")

        content = QWidget()
        content.setStyleSheet(f"background: {C.COLOR_1};")
        v = QVBoxLayout(content)
        v.setContentsMargins(0, 0, 0, 16)
        v.setSpacing(16)

        self._banner = Banner()
        v.addWidget(self._banner)

        self._title = QLabel("")
        self._title.setFont(QFont(UI_FONT, 26, QFont.Weight.Bold))
        self._title.setStyleSheet(f"color: {C.TEXT}; background: transparent;")
        self._title.setWordWrap(True)
        v.addWidget(self._title)

        info_row = QHBoxLayout()
        info_row.setSpacing(12)

        self._play_btn = QPushButton()
        self._play_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        self._play_btn.setFixedHeight(48)
        self._play_btn.setMinimumWidth(220)
        self._play_btn.setStyleSheet(_PLAY_QSS)
        play_lay = QHBoxLayout(self._play_btn)
        play_lay.setContentsMargins(20, 0, 16, 0)
        play_lay.setSpacing(5)
        self._play_icon_label = QLabel()
        self._play_icon_label.setPixmap(load_bundled_icon("play", "#ffffff").pixmap(20, 20))
        self._play_icon_label.setStyleSheet("background: transparent;")
        play_lay.addWidget(self._play_icon_label)
        self._play_text = QLabel("Play")
        self._play_text.setStyleSheet(
            f"background: transparent; color: #ffffff;"
            f"font-size: 17px; font-weight: 600;"
        )
        play_lay.addWidget(self._play_text)
        play_lay.addStretch(1)
        self._play_btn.clicked.connect(self._on_play)
        info_row.addWidget(self._play_btn)

        self._source_badge = QLabel("")
        self._source_badge.setStyleSheet(
            f"background-color: {C.COLOR_3}; color: {C.TEXT_DIM};"
            f"border-radius: {C.RADIUS}px; padding: 6px 12px; font-size: 12px;"
        )
        info_row.addWidget(self._source_badge, alignment=Qt.AlignmentFlag.AlignVCenter)

        self._path_label = QLabel("")
        self._path_label.setStyleSheet(f"color: {C.TEXT_DIM}; font-size: 12px; background: transparent;")
        self._path_label.setWordWrap(True)
        info_row.addWidget(self._path_label, stretch=1)

        v.addLayout(info_row)

        self._info_box = self._build_info_box()
        v.addWidget(self._info_box, alignment=Qt.AlignmentFlag.AlignLeft)

        v.addStretch(1)

        scroll.setWidget(content)
        outer.addWidget(scroll)

        self._banner.set_overlay(self._build_banner_overlay())
        self._banner_overlay = self._banner._overlay

    def _build_banner_overlay(self) -> QWidget:
        overlay = QWidget(self._banner)
        overlay.setStyleSheet("background: transparent;")

        lay = QVBoxLayout(overlay)
        lay.setContentsMargins(20, 16, 20, 20)
        lay.setSpacing(12)

        top = QHBoxLayout()
        back_btn = QPushButton("‹  Library")
        back_btn.setCursor(Qt.CursorShape.PointingHandCursor)
        back_btn.setStyleSheet(
            f"""
            QPushButton {{
                background-color: rgba(17, 17, 17, 170); color: {C.TEXT};
                border: none; border-radius: {C.RADIUS}px; padding: 8px 14px;
                font-size: 13px; font-weight: 600;
            }}
            QPushButton:hover {{ background-color: rgba(17, 17, 17, 230); }}
            """
        )
        back_btn.clicked.connect(self.back_requested)
        top.addWidget(back_btn)
        top.addStretch(1)
        lay.addLayout(top)

        lay.addStretch(1)

        self._logo_label = QLabel()
        self._logo_label.setAlignment(Qt.AlignmentFlag.AlignBottom | Qt.AlignmentFlag.AlignLeft)
        self._logo_label.setMaximumWidth(520)
        self._logo_label.setMinimumHeight(90)
        lay.addWidget(self._logo_label)

        return overlay

    def _build_info_box(self) -> QFrame:
        box = QFrame()
        box.setFixedWidth(300)
        box.setStyleSheet(
            f"QFrame {{ background-color: {C.COLOR_2}; border: none;"
            f"border-radius: {C.RADIUS}px; }}"
        )
        v = QVBoxLayout(box)
        v.setContentsMargins(14, 12, 14, 12)
        v.setSpacing(10)

        header = QLabel("GAME INFO")
        header.setStyleSheet(f"color: {C.TEXT_DIM}; font-size: 11px; font-weight: 700; letter-spacing: 1px;")
        v.addWidget(header)

        self._info_rows: dict[str, QLabel] = {}
        for key, label in (("Source", "source"), ("App ID", "app_id")):
            row = QHBoxLayout()
            row.setSpacing(8)
            k = QLabel(key.upper())
            k.setStyleSheet(f"color: {C.TEXT_DIM}; font-size: 11px;")
            k.setFixedWidth(70)
            val = QLabel("")
            val.setStyleSheet(f"color: {C.TEXT}; font-size: 12px; background: transparent;")
            val.setWordWrap(True)
            row.addWidget(k)
            row.addWidget(val, stretch=1)
            v.addLayout(row)
            self._info_rows[label] = val

        v.addStretch(1)
        return box

    def _on_play(self):
        if self.game is None:
            return
        if self._running:
            self.stop.emit(self.game)
        else:
            self.play.emit(self.game)

    def set_game(self, game: Game):
        self.game = game

        hero = art.load_hero(game, allow_network=False)
        if hero is None:
            hero = art.placeholder_card(game, 900, 420)
            self._banner.set_source(hero, fit=True)
        else:
            self._banner.set_source(hero)

        self._logo = art.load_logo(game)
        if self._logo is not None:
            self._logo_label.setPixmap(
                self._logo.scaledToWidth(
                    min(520, self._logo.width()),
                    Qt.TransformationMode.SmoothTransformation,
                )
            )
            self._title.setText("")
        else:
            self._logo_label.clear()
            self._title.setText(game.name.replace("/", " / "))

        self._source_badge.setText(game.source_name)
        self._path_label.setText(game.display_path)
        self._info_rows["source"].setText(game.source_name)
        self._info_rows["app_id"].setText(game.app_id or "—")

    def set_running(self, running: bool) -> None:
        if running == self._running:
            return
        self._running = running
        if running:
            self._play_btn.setStyleSheet(_RUNNING_QSS)
            self._play_icon_label.setPixmap(load_bundled_icon("stop", "#ffffff").pixmap(20, 20))
            self._play_text.setText("Stop")
        else:
            self._play_btn.setStyleSheet(_PLAY_QSS)
            self._play_icon_label.setPixmap(load_bundled_icon("play", "#ffffff").pixmap(20, 20))
            self._play_text.setText("Play")

    def set_hero(self, pix: QPixmap | None):
        if pix is None:
            return
        self._banner.set_source(pix)

File: src/ui/pages/GamePage.cpp
#include "ui/pages/GamePage.h"
#include "ui/widgets/Banner.h"
#include "ui/Theme.h"
#include "ui/Style.h"
#include "ui/Icons.h"
#include "ui/Fonts.h"
#include <QScrollBar>
#include <QFont>

using namespace theme::C;

static constexpr int BANNER_H = 420;

GamePage::GamePage(QWidget* parent)
    : QWidget(parent)
{
    style::background(this, COLOR_1);

    auto* outer = style::vbox(this, 0);
    outer->setContentsMargins(16, 16, 16, 16);

    auto* scroll = new QScrollArea(this);
    scroll->setWidgetResizable(true);
    scroll->setFrameShape(QFrame::NoFrame);
    scroll->setStyleSheet(QStringLiteral("QScrollArea { background-color: %1; border: none; }").arg(COLOR_1));

    auto* content = new QWidget();
    style::background(content, COLOR_1);
    auto* v = new QVBoxLayout(content);
    v->setContentsMargins(0, 0, 0, 16);
    v->setSpacing(16);

    m_banner = new Banner(content);
    v->addWidget(m_banner);

    m_title = style::heading("", 26, content);
    m_title->setWordWrap(true);
    v->addWidget(m_title);

    auto* infoRow = new QHBoxLayout();
    infoRow->setSpacing(12);

    m_playBtn = style::button(QString(), "play");
    m_playBtn->setFixedHeight(48);
    m_playBtn->setMinimumWidth(220);
    auto* playLay = new QHBoxLayout(m_playBtn);
    playLay->setContentsMargins(20, 0, 16, 0);
    playLay->setSpacing(5);
    m_playIconLabel = new QLabel();
    m_playIconLabel->setPixmap(icons::loadIcon("play", "#ffffff").pixmap(20, 20));
    style::transparent(m_playIconLabel);
    playLay->addWidget(m_playIconLabel);
    m_playText = new QLabel("Play");
    style::label(m_playText, "#ffffff", 17, 600);
    playLay->addWidget(m_playText);
    playLay->addStretch(1);
    connect(m_playBtn, &QPushButton::clicked, this, &GamePage::onPlayClicked);
    infoRow->addWidget(m_playBtn);

    m_sourceBadge = new QLabel();
    style::label(m_sourceBadge, TEXT_DIM, 12, -1,
        QStringLiteral("background-color: %1; padding: 6px 12px; border-radius: %2px;").arg(COLOR_3).arg(RADIUS));
    infoRow->addWidget(m_sourceBadge, 0, Qt::AlignVCenter);

    m_pathLabel = new QLabel();
    style::label(m_pathLabel, TEXT_DIM, 12);
    m_pathLabel->setWordWrap(true);
    infoRow->addWidget(m_pathLabel, 1);

    v->addLayout(infoRow);

    m_infoBox = buildInfoBox();
    v->addWidget(m_infoBox, 0, Qt::AlignLeft);

    m_achFrame = buildAchievementsBox();
    m_achFrame->hide();
    v->addWidget(m_achFrame);

    v->addStretch(1);

    scroll->setWidget(content);
    outer->addWidget(scroll);

    m_banner->setOverlay(buildBannerOverlay());
}

QWidget* GamePage::buildBannerOverlay() {
    auto* overlay = new QWidget(m_banner);
    style::transparent(overlay);

    auto* lay = new QVBoxLayout(overlay);
    lay->setContentsMargins(20, 16, 20, 20);
    lay->setSpacing(12);

    auto* top = new QHBoxLayout();
    auto* backBtn = style::button("‹  Library", "overlay");
    connect(backBtn, &QPushButton::clicked, this, &GamePage::backRequested);
    top->addWidget(backBtn);
    top->addStretch(1);
    lay->addLayout(top);

    lay->addStretch(1);

    m_logoLabel = new QLabel();
    m_logoLabel->setAlignment(Qt::AlignBottom | Qt::AlignLeft);
    m_logoLabel->setMaximumWidth(520);
    m_logoLabel->setMinimumHeight(90);
    lay->addWidget(m_logoLabel);

    return overlay;
}

QFrame* GamePage::makeBox(const QString& title, QVBoxLayout** outLay, QLabel** outHeader) {
    auto* box = style::card(nullptr, 2, 14, 12, 10);
    auto* v = static_cast<QVBoxLayout*>(box->layout());
    auto* header = new QLabel(title);
    style::label(header, TEXT_DIM, 11, 700);
    v->addWidget(header);
    if (outLay) *outLay = v;
    if (outHeader) *outHeader = header;
    return box;
}

QFrame* GamePage::buildInfoBox() {
    QVBoxLayout* v = nullptr;
    auto* box = makeBox("GAME INFO", &v);
    box->setFixedWidth(300);

    auto buildRow = [&](const QString& key, QLabel*& valLabel) {
        auto* row = new QHBoxLayout();
        row->setSpacing(8);
        auto* k = new QLabel(key.toUpper());
        style::label(k, TEXT_DIM, 11);
        k->setFixedWidth(70);
        valLabel = new QLabel();
        style::label(valLabel, TEXT, 12);
        valLabel->setWordWrap(true);
        row->addWidget(k);
        row->addWidget(valLabel, 1);
        v->addLayout(row);
    };

    buildRow("Source", m_infoSource);
    buildRow("App ID", m_infoAppId);

    v->addStretch(1);
    return box;
}

QFrame* GamePage::buildAchievementsBox() {
    QVBoxLayout* v = nullptr;
    auto* box = makeBox("ACHIEVEMENTS", &v, &m_achHeader);

    m_achList = new QListWidget();
    m_achList->setStyleSheet(style::listQss());
    m_achList->setMaximumHeight(180);
    v->addWidget(m_achList);
    return box;
}

void GamePage::populateAchievements(const Game& game) {
    Q_UNUSED(game);
    m_achFrame->hide();
}

void GamePage::setGame(const Game& game) {
    m_game = game;
    m_running = false;

    if (m_logo.isNull()) {
        m_title->setText(QString(game.name()).replace("/", " / "));
        m_logoLabel->clear();
    }

    m_sourceBadge->setText(game.sourceName());
    m_pathLabel->setText(game.displayPath());
    m_infoSource->setText(game.sourceName());
    m_infoAppId->setText(game.appId().isEmpty() ? "\u2014" : game.appId());

    updatePlayButton();
    populateAchievements(game);
}

void GamePage::setRunning(bool running) {
    if (m_game.name().isEmpty()) return;
    if (!m_game.isInstalled()) return;
    if (running == m_running) return;
    m_running = running;
    updatePlayButton();
}

void GamePage::setHero(const QPixmap& pix) {
    if (pix.isNull()) return;
    m_banner->setSource(pix);
}

void GamePage::updatePlayButton() {
    bool installed = m_game.isInstalled() && !m_game.path().isEmpty();
    if (installed) {
        m_playBtn->setStyleSheet(style::buttonQss(m_running ? "running" : "play"));
        m_playIconLabel->setPixmap(icons::loadIcon(m_running ? "stop" : "play", "#ffffff").pixmap(20, 20));
        m_playText->setText(m_running ? "Stop" : "Play");
    } else {
        m_playBtn->setStyleSheet(style::buttonQss("play"));
        m_playIconLabel->setPixmap(icons::loadIcon("box", "#ffffff").pixmap(20, 20));
        m_playText->setText("Install");
    }
}

void GamePage::onPlayClicked() {
    if (m_game.name().isEmpty()) return;
    if (!m_game.isInstalled() || m_game.path().isEmpty()) {
        emit install(m_game);
        return;
    }
    if (m_running) {
        emit stop(m_game);
    } else {
        emit play(m_game);
    }
}
